<a href="https://colab.research.google.com/github/Ratafar22/FoodHub_Analysis/blob/main/Foodhub_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Python Foundations: FoodHub Data Analysis

### Context

The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.

The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.

### Objective

The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are hired as a Data Scientist in this company and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business.

### Data Description

The data contains the different data related to a food order. The detailed data dictionary is given below.

### Data Dictionary

* order_id: Unique ID of the order
* customer_id: ID of the customer who ordered the food
* restaurant_name: Name of the restaurant
* cuisine_type: Cuisine ordered by the customer
* cost: Cost of the order
* day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
* rating: Rating given by the customer out of 5
* food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
* delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

### Let us start by importing the required libraries

In [ ]:
# import libraries for data manipulation
import numpy as np
import pandas as pd

# import libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

### Understanding the structure of the data

In [ ]:
# read the data
df = pd.read_csv('foodhub_order.csv')
# returns the first 5 rows
df.head()

order_id  customer_id            restaurant_name cuisine_type  \
0   1477147       337525                    Hangawi       Korean   
1   1477685       358141  Blue Ribbon Sushi Izakaya     Japanese   
2   1477070        66393                Cafe Habana      Mexican   
3   1477334       106968  Blue Ribbon Fried Chicken     American   
4   1478249        76942           Dirty Bird to Go     American   

   cost_of_the_order day_of_the_week     rating  food_preparation_time  \
0              30.75         Weekend  Not given                     25   
1              12.08         Weekend  Not given                     25   
2              12.23         Weekday          5                     23   
3              29.20         Weekend          3                     25   
4              11.59         Weekday          4                     25   

   delivery_time  
0             20  
1             23  
2             28  
3             15  
4             24

#### Observations:

The DataFrame has 9 columns as mentioned in the Data Dictionary. Data in each row corresponds to the order placed by a customer.

### **Question 1:** Write the code to check the shape of the dataset and write your observations based on that.

In [ ]:
# Write your code here
df.shape

(1898, 9)

#### Observations:
The dataset contains 9 columns and 1898 records.

### Question 2: Write the observations based on the below output from the info() method.

In [ ]:
# Use info() to print a concise summary of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               1898 non-null   int64  
 1   customer_id            1898 non-null   int64  
 2   restaurant_name        1898 non-null   object 
 3   cuisine_type           1898 non-null   object 
 4   cost_of_the_order      1898 non-null   float64
 5   day_of_the_week        1898 non-null   object 
 6   rating                 1898 non-null   object 
 7   food_preparation_time  1898 non-null   int64  
 8   delivery_time          1898 non-null   int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 133.6+ KB


#### Observations:
All the columns in the dataset have no missing values and have the same number of rows (1898). Order_id, customer_id, food preparation time and delivery time are of the datatype int64. restaurant_name, cusine_type, day_of_the_week and rating are object datatype while only cost_of_the_order is a float data type.

### Question 3: 'restaurant_name', 'cuisine_type', 'day_of_the_week' are object types. Write the code to convert the mentioned features to 'category' and write your observations on the same.

In [ ]:
# Coverting "objects" to "category" reduces the data space required to store the dataframe
# write the code to convert 'restaurant_name', 'cuisine_type', 'day_of_the_week' into categorical data
df = df.astype({'restaurant_name':'category', 'cuisine_type':'category', 'day_of_the_week':'category'})

# Use info() to print a concise summary of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   order_id               1898 non-null   int64   
 1   customer_id            1898 non-null   int64   
 2   restaurant_name        1898 non-null   category
 3   cuisine_type           1898 non-null   category
 4   cost_of_the_order      1898 non-null   float64 
 5   day_of_the_week        1898 non-null   category
 6   rating                 1898 non-null   object  
 7   food_preparation_time  1898 non-null   int64   
 8   delivery_time          1898 non-null   int64   
dtypes: category(3), float64(1), int64(4), object(1)
memory usage: 102.7+ KB




```
# This is formatted as code
```

#### Observations:
The dataset now has 3 category, 1 float64, 4 int64, and 1 object datatypes, also the memory usage has decreased from 133.6 KB to 102.7 KB. The other information about the dataset remains the same.

### **Question 4:** Write the code to find the summary statistics and write your observations based on that.

In [ ]:
# Write your code here
df.describe()

order_id    customer_id  cost_of_the_order  food_preparation_time  \
count  1.898000e+03    1898.000000        1898.000000            1898.000000   
mean   1.477496e+06  171168.478398          16.498851              27.371970   
std    5.480497e+02  113698.139743           7.483812               4.632481   
min    1.476547e+06    1311.000000           4.470000              20.000000   
25%    1.477021e+06   77787.750000          12.080000              23.000000   
50%    1.477496e+06  128600.000000          14.140000              27.000000   
75%    1.477970e+06  270525.000000          22.297500              31.000000   
max    1.478444e+06  405334.000000          35.410000              35.000000   

       delivery_time  
count    1898.000000  
mean       24.161749  
std         4.972637  
min        15.000000  
25%        20.000000  
50%        25.000000  
75%        28.000000  
max        33.000000

#### Observations:


*   The cost computes the summary statistics for columns with Int64 data type
*   Order ID and Customer ID are simply identifiers for orders and customers.
* The average order cost is 16.5 dollars, with a minimum of 4.5 dollars and a maximum of $35.5.
*   Food preparation time ranges from 20 to 35 minites, with an avrage of 27.4 minutes.
*  Delivery time ranges from a minimum of 15 minutes to a maximum of 33 minutes, with an average delivery time of 24.2 minutes.





### **Question 5:** How many orders are not rated?

In [ ]:
df.groupby('rating').rating.count()

rating
3            188
4            386
5            588
Not given    736
Name: rating, dtype: int64

#### Observations:
There are 736 orders with no rating

### Exploratory Data Analysis (EDA)




### Question 6: Write the code to find the top 5 restaurants that have received the highest number of orders.

In [ ]:
#df.groupby('restaurant_name').order_id.count().sort_values(by =count_orders ascending=False)
df.groupby('restaurant_name').order_id.count().sort_values( ascending=False).head()

<ipython-input-8-3d08fddd2815>:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('restaurant_name').order_id.count().sort_values( ascending=False).head()


restaurant_name
Shake Shack                  219
The Meatball Shop            132
Blue Ribbon Sushi            119
Blue Ribbon Fried Chicken     96
Parm                          68
Name: order_id, dtype: int64

#### Observations:
The top 5 restaurants with the highest number of orders are Shake Schack with 219 orders, followed by The Meatball Shop with 132, Blue Ribbon Sushi having 119 orders, Blue Ribbon Fried Chicken with 96 and Parm with 68 orders.

These orders constitute about 44% of the total food ordered from all restaurants.



### Question 7: Write the code to find the most popular cuisine on weekends.

In [ ]:
# define a df for weekend
weekend_orders = df[df['day_of_the_week']== 'Weekend']

# find the most popular cuisine type ordered on weekends
weekend_orders.cuisine_type.value_counts().head(1)


cuisine_type
American    415
Name: count, dtype: int64

#### Observations:
American is the most popular cuisine type ordered on weekends with a total of 415 orders.

### Question 8: Write the code to find the number of total orders where the cost is above 20 dollars. What is the percentage of such orders in the dataset?

In [ ]:
# number of orders with cost above 20 dollars
df[df['cost_of_the_order'] > 20]['cost_of_the_order'].value_counts().sum()

555

In [ ]:
#percentage of the orders in the dataset

df[df['cost_of_the_order'] > 20]['cost_of_the_order'].value_counts().sum()/df.shape[0]*100

29.24130663856691

#### Observations:
There are 555 orders with a cost more than 20 dollars. Such orders accounted for 29.2% of the total orders in the dataset. This means that about 80% of the orders cost less than 20 dollars.

### Question 9: Write the code to find the mean delivery time based on this dataset. (1 mark)

In [ ]:
# Write the code here
round(df.delivery_time.mean(),1)

24.2

#### Observations:
On average, it takes 24.2 minutes to deliver an order

### Question 10: Suppose the company has decided to give a free coupon of 15 dollars to the customer who has spent the maximum amount on a single order. Write the code to find the ID of the customer along with the order details.

In [ ]:
max_order =df.loc[df['cost_of_the_order'].idxmax()]
max_order

order_id                       1477814
customer_id                      62359
restaurant_name                  Pylos
cuisine_type             Mediterranean
cost_of_the_order                35.41
day_of_the_week                Weekday
rating                               4
food_preparation_time               21
delivery_time                       29
Name: 573, dtype: object

#### Observations:
Customer with ID 62359 spent the highest amount on a single order.





### Question 11: Suppose the company wants to provide a promotional offer in the advertisement of the restaurants. The condition to get the offer is that the restaurants must have a rating count of more than 50 and the average rating should be greater than 4. Write the code to find the restaurants fulfilling the criteria to get the promotional offer.

In [ ]:
#create a dataframe for the orders with ratings
order_with_ratings = df.loc[df['rating'] != 'Not given']

# change the datatype of the rating to integer
order_with_ratings = order_with_ratings.astype({'rating':'int64'})

#order_with_ratings.info()

# rating count and average rating for each restaurant
order_with_ratings = order_with_ratings.groupby('restaurant_name').rating.agg(['count', 'mean'])\
.rename(columns = {'count': 'rating_count', 'mean':'average_ratings'})\
.sort_values(['rating_count'], ascending= False).reset_index()

# eligible restaurants
eligible_restaurants = order_with_ratings[(order_with_ratings['rating_count'] > 50) & (order_with_ratings['average_ratings'] > 4)]
eligible_restaurants

<ipython-input-14-2c4714f6d5dc>:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  order_with_ratings = order_with_ratings.groupby('restaurant_name').rating.agg(['count', 'mean'])\


restaurant_name  rating_count  average_ratings
0                Shake Shack           133         4.278195
1          The Meatball Shop            84         4.511905
2          Blue Ribbon Sushi            73         4.219178
3  Blue Ribbon Fried Chicken            64         4.328125

#### Observations:
The four restaurants that met the criteria for the promotional offer are Shake Shack, The Meatball Shop, Blue Ribbon and Blue Ribbon Fried Chicken.

### Question 12: Suppose the company charges the restaurant 25% on the orders having cost greater than 20 dollars and 15% on the orders having cost greater than 5 dollars. Write the code to find the net revenue generated on all the orders given in the dataset.

In [ ]:
# Total cost
df['cost_of_the_order'].sum()

# define a  function to calculate the commision on the order cost
def revenue_commision(cost):
  if cost > 20:
    return cost * 0.25
  elif cost> 5 and cost <=20:
    return cost * 0.15
  else:
    return 0

#Use the function to calculate the commission for each order
commission = df['cost_of_the_order'].apply(revenue_commision)


#Calculate the total revenue
net_revenue = commission.sum()
print(net_revenue)
print(df['cost_of_the_order'].sum())

6166.303
31314.82


#### Observations:
Assumming the company wants to charge fees based on the specified cost_of_the_order, they might make a net revenue of $6,166.3 which is about 20% of the order_cost.

### Question 13: Suppose the company wants to analyze the total time required to deliver the food. Write the code to find out the percentage of orders that have more than 60 minutes of total delivery time. (2 marks)

Note: The total delivery time is the summation of the food preparation time and delivery time.

In [ ]:
# Sum the food_prepation_time and delivery time columns
prep_time = (df.food_preparation_time + df.delivery_time).rename('total_delivery_time')
prep_time

#filter for the delivery time greater than 60 minutes
prep_time_greater_than_60mins = prep_time[prep_time > 60].value_counts().sum()
print(prep_time_greater_than_60mins)

# calculate the percentage of these orders
percent_of_the_orders = round(prep_time_greater_than_60mins/df.shape[0]*100,1)
print(percent_of_the_orders)

200
10.5


#### Observations:

A total of 200 orders takes more 60 minutes to be delivered. These orders represent 10.5% of all orders placed from the restaurants.

### Question 14: Suppose the company wants to analyze the delivery time of the orders on weekdays and weekends. Write the code to find the mean delivery time on weekdays and weekends. Write your observations on the results. (2 marks)

In [ ]:
df.groupby('day_of_the_week').delivery_time.mean()

<ipython-input-17-ec9965e8f2fe>:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day_of_the_week').delivery_time.mean()


day_of_the_week
Weekday    28.340037
Weekend    22.470022
Name: delivery_time, dtype: float64

#### Observations:

Delivery during the weekdays takes, on average, about 6 minutes longer than delivery done during the weekends.

### Conclusion and Recommendations

### **Question 15:** Write the conclusions and business recommendations derived from the analysis. (3 marks)

#### **Key Insights**:


*   The top 5 restaurants with the highest number of orders are Shake Shack, Meatball Shop, Blue Ribbon Sushi, Blue Ribbon Fried Chicken, and Parm, accounting for 44% of total orders. Among them, only Parm received fewer than 50 ratings counts, despite having an average rating above 4.

*  American, Japanese, and Italian cuisines are the most popular choices on both weekdays and weekends.

*  Approximately 80% of all orders cost less than $20 each.


*  38.8% of orders were not rated by customers.

* The average delivery time is 24.2 minutes. Interestingly, even though 71% of orders were placed on weekends, deliveries were faster compared to weekdays. This could be due to:
  - Reduced weekday traffic congestion, making deliveries slower during workdays.
  - More delivery drivers available on weekends, improving efficiency.
  - Orders being placed from home rather than offices, leading to shorter delivery distances.

*  The company could generate a net revenue of 6,166.30 dollars (~20% of total order cost) if it charges restaurants 25% on orders above 20 dollars and 15% on orders above $5.



**Recommedations**


*   Reward high-performing restaurants: Offer incentives to restaurants that consistently receive high orders and positive ratings to motivate quality service and encourage other restaurants to improve.

*  Encourage customer ratings: Since 38.8% of orders remain unrated, the platform should redesign the rating system to be more engaging and User-friendly, possibly through interactive prompts or small incentives for leaving a review.

*   Optimize delivery times: Although only ~11% of orders exceed 60 minutes, the company should further reduce wait times to improve customer satisfaction and retention. This can be achieved by:
  - Strategic driver allocation during peak hours.
  - Partnering with more delivery personnel to meet demand.
  - Improving restaurant workflow efficiency to shorten preparation time.








